dpnp/backend/src/queue_sycl.hpp

```diff
@@ -40,25 +41,108 @@
 
 #include <ctime>
 
-#if !defined(DPNP_LOCAL_QUEUE)
-#if defined __has_include
-#if __has_include(<dpctl_sycl_interface.h>)
-#include <dpctl_sycl_interface.h>
-#else
-#include <dpctl_sycl_queue_manager.h>
-#endif
-#endif
-#endif
-
 #include "dpnp_pstl.hpp" // this header must be included after <mkl.hpp>
 
 #include "verbose.hpp"
 
 namespace mkl_rng = oneapi::mkl::rng;
 
-#define DPNP_QUEUE            backend_sycl::get_queue()
-#define DPNP_RNG_ENGINE       backend_sycl::get_rng_engine()
-#define DPNP_RNG_MCG59_ENGINE backend_sycl::get_rng_mcg59_engine()
+#define DPNP_QUEUE            backend_sycl_singleton::get_queue()
+#define DPNP_RNG_ENGINE       backend_sycl_singleton::get_rng_engine()
+#define DPNP_RNG_MCG59_ENGINE backend_sycl_singleton::get_rng_mcg59_engine()
+
+class backend_sycl_singleton {
+public:
+    ~backend_sycl_singleton() {}
+
+    static backend_sycl_singleton& get() {
+        static backend_sycl_singleton backend = lookup();
+        return backend;
+    }
+
+    static sycl::queue& get_queue() {
+        auto &be = backend_sycl_singleton::get();
+        return *(be.queue_ptr);
+    } 
+
+    static mkl_rng::mt19937& get_rng_engine() {
+        auto &be = backend_sycl_singleton::get();
+        return *(be.rng_mt19937_engine_ptr);
+    } 
+
+    static mkl_rng::mcg59& get_rng_mcg59_engine() {
+        auto &be = backend_sycl_singleton::get();
+        return *(be.rng_mcg59_engine_ptr);
+    }
+
+    template <typename SeedT>
+    void set_rng_engines_seed(const SeedT &seed) {
+        auto rng_eng_mt19937 = std::make_shared<mkl_rng::mt19937>(*queue_ptr, seed);
+        if (!rng_eng_mt19937) {
+            throw std::runtime_error(
+                "Could not create MT19937 engine with given seed"
+            );
+        }
+        auto rng_eng_mcg59 = std::make_shared<mkl_rng::mcg59>(*queue_ptr, seed);
+        if (!rng_eng_mcg59) {
+            throw std::runtime_error(
+                "Could not create MCG59 engine with given seed"
+            );
+        }
+
+        rng_mt19937_engine_ptr.swap(rng_eng_mt19937);
+        rng_mcg59_engine_ptr.swap(rng_eng_mcg59);
+    }
+
+    bool backend_sycl_is_cpu() const {
+        if (queue_ptr) {
```

Comment:
Is it possible to have `queue_ptr == nullptr`? If memory can't be allocated for `queue_ptr`, then there will be runtime exception thrown from the constructor.
Otherwise, probably we also need to add similar check `if (queue_ptr)` to `set_rng_engines_seed` method.

Reply:
The `shared_ptr<queue>` is default initialized by default constructor and so has `nullptr` content until after it is reset with `std::make_shared`. This command would call `sycl::queue` constructor which may throw an exception is queue construction was not successful. So exception will be thrown, but I'd keep the check here.

Reply:
Actually, since constructor of the class is private, it can only be created via `get` static method, which would call private constructor. Any exceptions raised during this constructor call would result in the instance of this class not being created, and hence all `bool(queue_ptr)` tests in regular methods are superfluous. I am going to remove them.